programs/fileio.c

```diff
@@ -2147,6 +2244,82 @@ FIO_fwriteSparseEnd(const FIO_prefs_t* const prefs, FILE* file, unsigned storedS
     }   }
 }
 
+static void FIO_writePoolSetDstFile(write_pool_ctx_t *ctx, FILE* dstFile) {
+    assert(ctx!=NULL);
+    // We can change the dst file only if we have finished writing
+    if(ctx->writerPool)
+        POOL_joinJobs(ctx->writerPool);
+    assert(ctx->storedSkips == 0);
+    assert(ctx->availableWriteJobs == ctx->totalWriteJobs);
+    ctx->dstFile = dstFile;
+}
+
+static int FIO_writePoolCloseDstFile(write_pool_ctx_t *ctx) {
+    FILE *dstFile = ctx->dstFile;
+    assert(dstFile!=NULL || ctx->prefs->testMode!=0);
+    FIO_writePoolSetDstFile(ctx, NULL);
+    return fclose(dstFile);
+}
+
+
+static void FIO_releaseWriteJob(write_job_t *job) {
+    write_pool_ctx_t *ctx = job->ctx;
+    if(ctx->writerPool) {
+        ZSTD_pthread_mutex_lock(&ctx->writeJobsMutex);
+        assert(ctx->availableWriteJobs < DECOMPRESSION_MAX_WRITE_JOBS);
+        ctx->jobs[ctx->availableWriteJobs++] = job;
+        ZSTD_pthread_mutex_unlock(&ctx->writeJobsMutex);
+    } else {
+        ctx->availableWriteJobs++;
+    }
+}
+
+static write_job_t* FIO_writePoolGetAvailableWriteJob(write_pool_ctx_t *ctx) {
+    write_job_t *job;
+    assert(ctx->dstFile!=NULL || ctx->prefs->testMode);
+    if(ctx->writerPool) {
+        ZSTD_pthread_mutex_lock(&ctx->writeJobsMutex);
+        assert(ctx->availableWriteJobs > 0);
+        job = (write_job_t*) ctx->jobs[--ctx->availableWriteJobs];
+        ZSTD_pthread_mutex_unlock(&ctx->writeJobsMutex);
+    } else {
+        assert(ctx->availableWriteJobs==1);
+        ctx->availableWriteJobs--;
+        job = (write_job_t*)ctx->jobs[0];
+    }
+    job->usedBufferSize = 0;
+    job->dstFile = ctx->dstFile;
+    return job;
+}
+
+static void FIO_WritePoolWriteJobExecute(void* opaque){
+    write_job_t* job = (write_job_t*) opaque;
+    write_pool_ctx_t* ctx = job->ctx;
+    ctx->storedSkips = FIO_fwriteSparse(job->dstFile, job->buffer, job->usedBufferSize, ctx->prefs, ctx->storedSkips);
+    FIO_releaseWriteJob(job);
+}
+
+static void FIO_writePoolQueueWriteJob(write_job_t *job) {
+    write_pool_ctx_t* ctx = job->ctx;
+    if(ctx->writerPool)
+        POOL_add(ctx->writerPool, FIO_WritePoolWriteJobExecute, job);
+    else
+        FIO_WritePoolWriteJobExecute(job);
+}
+
+static void FIO_writePoolQueueWriteJobAndGetNextAvailable(write_job_t **job) {
```

Comment:
For all the functions you intend to be "public", e.g. not the helper functions only used by the other `WritePool` functions, could you add documentation? You could also add docs for the helpers, but that is less necessary.

Reply:
Fixed